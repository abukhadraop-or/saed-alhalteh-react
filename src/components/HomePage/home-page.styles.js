import { Link } from 'react-router-dom';
import styled from 'styled-components';
import breakingPoints from 'styles/breaking-points';
import colors from 'styles/colors';

export const BannerContainer = styled.div`
  background-color: ${colors.green};
  box-shadow: inset 0 0.5rem 0.5rem -0.5rem ${colors.transparentBlack},
    inset 0 -0.5rem 0.5rem -0.5rem ${colors.transparentBlack};
  color: ${colors.white};
  padding: 1.55rem;
  text-align: center;
`;

export const BannerTitle = styled.h1`
  font-size: 3rem;
  margin: 0 0 1.55rem;
  text-shadow: 0 0.06rem 0.18rem ${colors.black};
`;

export const BannerDescription = styled.div`
  font-size: 1.5rem;
  font-weight: 300;
  margin: 0;
`;

export const NavItemsContainer = styled.ul`
  border-bottom: 0.125rem solid ${colors.concrete};
  list-style-type: none;
  margin: 3.125rem 0 1.563rem 0;
  min-height: 1.875rem;
  padding: 0;
  position: relative;

  @media (min-width: ${breakingPoints.mobile}) and (max-width: ${breakingPoints.tablets}) {
    margin: 3.125rem 0 1.5625rem 0;
  }

  @media (min-width: ${breakingPoints.tablets}) {
    margin: 3.125rem 2.5rem 1.5625rem 2.5rem;
  }

  @media (min-width: ${breakingPoints.largeScreen}) {
    margin: 3.125rem 6.25rem 1.5625rem 6.25rem;
  }
`;

export const NavItem = styled.li`
  display: inline-block;
`;

export const NavLink = styled(Link)`
  border-bottom: 0.125rem solid ${colors.green};
  bottom: -0.125rem;
  color: ${colors.green};
  display: inline-block;
  margin: 0;
  padding: 0.5rem 1rem;
  position: absolute;
  text-decoration: none;
`;

export const TagsAndPostsContainer = styled.div`
  @media (min-width: ${breakingPoints.tablets}) {
    display: flex;
    flex-direction: row-reverse;
    justify-content: center;
  }

  @media (min-width: ${breakingPoints.largeScreen}) {
    justify-content: space-between;
  }
`;

export const PostsDiv = styled.div`
  flex-grow: 1;
`;
