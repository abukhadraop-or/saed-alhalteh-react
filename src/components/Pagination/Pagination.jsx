import React from 'react';
import PropTypes from 'prop-types';
import { PaginationList, PaginationLink } from './pagination.styles';

/**
 * Component that displays the pages number based on the number of available posts.
 *
 * @param {Array} posts        Array of objects, which contains the posts.
 * @param {Number} pageSize    Maximum number of posts that are allowed per page.
 * @param {function} onClick   functions to set the range of posts to be displayed based on the current page.
 * @param {Number} CurrentPage current page.
 *
 * @returns {JSX.Element} Pagination
 */

function Pagination({ posts, pageSize, onClick, currentPage }) {
  const itemCount = posts.length;
  let pagesCount = Math.floor(itemCount / pageSize) + 1;
  if (itemCount % pageSize === 0) pagesCount -= 1;
  if (pagesCount === 1) return null;
  const pagesArray = [];
  for (let i = 1; i <= pagesCount; i += 1) {
    pagesArray.push(i);
  }
  return (
    <div>
      <PaginationList currentPage={currentPage}>
        {pagesArray.map((page) => (
          <PaginationLink key={page} onClick={() => onClick(page)}>
            {page}
          </PaginationLink>
        ))}
      </PaginationList>
    </div>
  );
}
Pagination.propTypes = {
  pageSize: PropTypes.number.isRequired,
  onClick: PropTypes.func.isRequired,
  currentPage: PropTypes.number.isRequired,
  // eslint-disable-next-line react/forbid-prop-types
  posts: PropTypes.array.isRequired,
};
export default Pagination;
