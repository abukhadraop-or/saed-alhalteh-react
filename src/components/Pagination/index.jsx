import PropTypes from 'prop-types';
import React from 'react';
import { PaginationButton, PaginationItem, PaginationList } from './pagination.styles';

/**
 * Pagination.
 *
 * @param {Object} props The props object.
 * @param {number} props.count  Posts count.
 * @param {number} props.pageSize Maximum allowed number of posts per page.
 * @param {Function} props.onClick Handles clicks on pagination.
 * @param {Number} props.currentPage The current page.
 *
 * @return {JSX.Element} Pagination component.
 */
function Pagination({ count, pageSize, onClick, currentPage }) {
  let pagesCount = Math.floor(count / pageSize) + 1;
  pagesCount = count % pageSize === 0 ? pagesCount - 1 : pagesCount;

  const pagesArray = [];

  for (let i = 1; i <= pagesCount; i += 1) {
    pagesArray.push(i);
  }

  return pagesCount === 1 ? null : (
    <PaginationList currentPage={currentPage}>
      {pagesArray.map((page) => (
        <PaginationItem key={page}>
          <PaginationButton onClick={() => onClick(page)}>{page}</PaginationButton>
        </PaginationItem>
      ))}
    </PaginationList>
  );
}

Pagination.propTypes = {
  pageSize: PropTypes.number.isRequired,
  onClick: PropTypes.func.isRequired,
  currentPage: PropTypes.number.isRequired,
  count: PropTypes.number.isRequired,
};

export default Pagination;
