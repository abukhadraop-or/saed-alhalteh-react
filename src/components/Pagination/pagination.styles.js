import styled from 'styled-components';
import { colors } from 'shared/styles';

export const PaginationList = styled.ul`
  list-style: none;
  text-align: center;

  & :nth-child(${(props) => props.currentPage}) {
    background-color: ${colors.fern};
    color: ${colors.white};
  }
`;

export const PaginationLink = styled.li`
  border: 0.0625rem solid ${colors.fern};
  border-radius: 0.1875rem;
  color: ${colors.fern};
  cursor: pointer;
  display: inline-block;
  margin: 0 0.3125rem 0 0;
  padding: 0.4375rem;
`;
