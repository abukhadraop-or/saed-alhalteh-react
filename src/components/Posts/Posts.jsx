import MetaInfo from 'components/MetaInfo/MetaInfo';
import PropTypes from 'prop-types';
import React from 'react';
import {
  Container,
  Like,
  PostContent,
  ReadMoreDiv,
  ReadMoreLink,
  UserDataContainer,
} from './posts.styles';

/**
 * Posts details.
 *
 * @param {object} props The props object.
 * @param {object[]} props.posts Posts.
 *
 * @return {JSX.Element} Posts component.
 */
function Posts({ posts }) {
  return posts.length === 0 ? (
    <Container>
      <p> No articles are here... yet. </p>
    </Container>
  ) : (
    posts.map((post) => (
      <Container key={post.slug}>
        <UserDataContainer>
          <MetaInfo author={post.author} createdAt={post.createdAt} />
          <Like count={post.favoritesCount} />
        </UserDataContainer>
        <PostContent to={{ pathname: '/posts', state: { detail: post } }}>
          <h2>{post.title}</h2>
          <p>{post.description}</p>
        </PostContent>
        <ReadMoreDiv>
          <ReadMoreLink to={{ pathname: '/posts', state: { detail: post } }}>
            Read more...
          </ReadMoreLink>
        </ReadMoreDiv>
      </Container>
    ))
  );
}

Posts.defaultProps = {
  posts: [],
};

Posts.propTypes = {
  posts: PropTypes.arrayOf(
    PropTypes.shape({
      title: PropTypes.string,
      description: PropTypes.string,
      favoritesCount: PropTypes.number,
      createdAt: PropTypes.string,
      author: PropTypes.shape({
        username: PropTypes.string,
        bio: PropTypes.string,
        image: PropTypes.string,
        following: PropTypes.bool,
      }),
    })
  ),
};

export default Posts;
