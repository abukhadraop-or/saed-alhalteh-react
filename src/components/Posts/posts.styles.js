import LikeComponent from 'components/Like';
import { Link } from 'react-router-dom';
import styled from 'styled-components';
import breakingPoints from 'styles/breaking-points';
import colors from 'styles/colors';

export const Container = styled.div`
  padding: 0 0.625rem;

  @media (min-width: ${breakingPoints.mobile}) {
    margin: 0 auto;
  }

  @media (min-width: ${breakingPoints.tablets}) {
    margin: 0 0 0 2.5rem;
    padding: 0 2rem 0 0.625rem;
  }

  @media (min-width: ${breakingPoints.largeScreen}) {
    margin: 0 0 0 6.25rem;
  }
`;

export const UserDataContainer = styled.div`
  display: flex;
  padding: 0 1.25rem 0 0;
`;

export const Like = styled(LikeComponent)`
  & span {
    margin: 0;
  }

  @media (min-width: ${breakingPoints.tablets}) {
    margin: 0 0.9375rem 0 auto;
  }
`;

export const PostContent = styled(Link)`
  display: inline-block;
  text-decoration: none;

  & h2 {
    color: ${colors.black};
    font-size: 1.5rem;
    font-weight: 600;
    margin: 0.5rem 0;
  }

  & p {
    color: ${colors.dustyGrey};
    font-size: 1rem;
    font-weight: 300;
    margin: 0.25rem 0;
  }
`;

export const ReadMoreDiv = styled.div`
  padding: 0.5rem 0 1.5rem;
`;

export const ReadMoreLink = styled(Link)`
  color: ${colors.dustyGrey};
  cursor: pointer;
  font-size: 0.75rem;

  text-decoration: none;

  &:hover {
    text-decoration: underline;
  }
`;
