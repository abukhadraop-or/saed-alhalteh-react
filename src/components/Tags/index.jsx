import PropTypes from 'prop-types';
import React, { useEffect, useState } from 'react';
import getTags from 'services/fake-tag-lists';
import { Tag, TagsBox, TagsContainer, Title } from './tags.styles';

/**
 * Tags section.
 *
 * @param {Object} props The props object.
 * @param {Function} props.onClick Handles click events on tags.
 * @param {String[]} props.selectedTags The selected tags.
 *
 * @return {JSX.Element} Tags component.
 */
function Tags({ onClick, selectedTags }) {
  const [tags, setTags] = useState([]);

  useEffect(() => {
    const result = getTags();
    setTags(result.tags);
  }, []);

  return (
    <TagsContainer>
      <Title> Popular Tags</Title>
      <TagsBox>
        {tags.length === 0 ? (
          <b> No Tags are available yet... </b>
        ) : (
          tags.map((tag) => (
            <Tag key={tag} onClick={() => onClick(tag)} isClicked={selectedTags.includes(tag)}>
              {tag}
            </Tag>
          ))
        )}
      </TagsBox>
    </TagsContainer>
  );
}

Tags.propTypes = {
  onClick: PropTypes.func.isRequired,
  selectedTags: PropTypes.arrayOf(PropTypes.string).isRequired,
};

export default Tags;
