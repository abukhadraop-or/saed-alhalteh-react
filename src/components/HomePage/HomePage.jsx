import Pagination from 'components/Pagination/Pagination';
import Posts from 'components/Posts/Posts';
import Tags from 'components/Tags/Tags';
import React, { useEffect, useState } from 'react';
import getCurrentPagePosts from 'services/fake-posts-service';
import {
  BannerContainer,
  BannerDescription,
  BannerTitle,
  NavItem,
  NavItemsContainer,
  NavLink,
  PostsDiv,
  TagsAndPostsContainer,
} from './home-page.styles';

/**
 * @type {number} Maximum number of posts per page.
 */
const pageSize = 5;

/**
 * The home page of the application.
 *
 * @return {JSX.Element} HomePage component.
 */
function HomePage() {
  const [posts, setPosts] = useState([]);
  const [currentPage, setCurrentPage] = useState(1);
  const [selectedTags, setSelectedTags] = useState([]);
  const [postsCount, setPostsCount] = useState(0);

  useEffect(() => {
    const data = getCurrentPagePosts(currentPage, pageSize, selectedTags);
    setPosts(data.posts);
    setPostsCount(data.count);
  }, [currentPage, selectedTags]);

  /**
   * Handle clicking on a tag.
   *
   * @param {string} tag Clicked tag.
   */
  const handleTagClick = (tag) => {
    const tagsArray = selectedTags;

    if (tagsArray.indexOf(tag) !== -1) {
      tagsArray.splice(tagsArray.indexOf(tag), 1);
    } else {
      tagsArray.push(tag);
    }

    setSelectedTags([...tagsArray]);
    setCurrentPage(1);
  };

  return (
    <>
      <BannerContainer>
        <BannerTitle> Conduit </BannerTitle>
        <BannerDescription> A place to share your knowledge </BannerDescription>
      </BannerContainer>
      <NavItemsContainer>
        <NavItem>
          <NavLink to="/"> Global Feed</NavLink>
        </NavItem>
      </NavItemsContainer>
      <TagsAndPostsContainer>
        <Tags onClick={(tag) => handleTagClick(tag)} selectedTags={selectedTags} />
        <PostsDiv>
          <Posts posts={posts} />
        </PostsDiv>
      </TagsAndPostsContainer>
      <Pagination
        count={postsCount}
        pageSize={pageSize}
        currentPage={currentPage}
        onClick={(page) => setCurrentPage(page)}
      />
    </>
  );
}

export default HomePage;
