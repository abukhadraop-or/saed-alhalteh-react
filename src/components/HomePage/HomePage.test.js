import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import Pagination from 'components/Pagination/Pagination';
import { PaginationButton } from 'components/Pagination/pagination.styles';
import Posts from 'components/Posts/Posts';
import Tags from 'components/Tags/Tags';
import { Tag } from 'components/Tags/tags.styles';
import Enzyme, { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import getCurrentPagePosts from 'services/fake-posts-service';
import HomePage from './HomePage';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('services/fake-posts-service');

const mockData = {
  posts: [
    {
      id: 'gvnzew4eh6d87v3',
      slug: 'how-to-train-your-dragon',
      title: 'How to train your dragon',
      description: 'Ever wonder how?',
      body: '1 Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
      tagList: ['dragons', 'training', 'odd', 'only the first Article'],
      createdAt: '2016-02-18T03:22:56.637Z',
      updatedAt: '2016-02-18T03:48:35.824Z',
      favorited: false,
      favoritesCount: 27,
      author: {
        username: 'jake',
        bio: 'I work at statefarm',
        image:
          'https://upload.wikimedia.org/wikipedia/en/thumb/2/24/Crystal_personal.svg/1024px-Crystal_personal.svg.png',
        following: false,
      },
    },
  ],
  count: 1,
};

describe('<HomePage /> tests', () => {
  it('matches the snapshot', () => {
    const wrapper = shallow(<HomePage />);

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('passes posts and count correctly to Posts and Pagination components ', () => {
    getCurrentPagePosts.mockReturnValue(mockData);
    const mountWrapper = mount(
      <BrowserRouter>
        <HomePage />
      </BrowserRouter>
    );
    const { posts } = mountWrapper.find(Posts).props();
    const { count } = mountWrapper.find(Pagination).props();

    expect(posts).toBe(mockData.posts);
    expect(count).toBe(1);
  });

  describe('#handleTagClick', () => {
    it('sets selectedTags and current page correctly', () => {
      getCurrentPagePosts.mockReturnValue(mockData);
      const mountWrapper = mount(
        <BrowserRouter>
          <HomePage />
        </BrowserRouter>
      );

      mountWrapper.find(Tag).at(1).simulate('click');

      let { selectedTags } = mountWrapper.find(Tags).props();
      let { currentPage } = mountWrapper.find(Pagination).props();

      expect(selectedTags).toEqual(['even']);
      expect(currentPage).toEqual(1);

      mountWrapper.find(Tag).at(1).simulate('click');

      ({ selectedTags } = mountWrapper.find(Tags).props());
      ({ currentPage } = mountWrapper.find(Pagination).props());

      expect(selectedTags).not.toEqual(['even']);
      expect(currentPage).toEqual(1);
    });
  });

  describe('pagination onClick', () => {
    it('sets current page correctly', () => {
      const mockData2 = {
        posts: [
          {
            id: 'gvnzew4eh6d87v3',
            slug: 'how-to-train-your-dragon',
            title: 'How to train your dragon',
            description: 'Ever wonder how?',
            body: '1 Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
            tagList: ['dragons', 'training', 'odd', 'only the first Article'],
            createdAt: '2016-02-18T03:22:56.637Z',
            updatedAt: '2016-02-18T03:48:35.824Z',
            favorited: false,
            favoritesCount: 27,
            author: {
              username: 'jake',
              bio: 'I work at statefarm',
              image:
                'https://upload.wikimedia.org/wikipedia/en/thumb/2/24/Crystal_personal.svg/1024px-Crystal_personal.svg.png',
              following: false,
            },
          },
        ],
        count: 12,
      };
      getCurrentPagePosts.mockReturnValue(mockData2);
      const mountWrapper = mount(
        <BrowserRouter>
          <HomePage />
        </BrowserRouter>
      );

      mountWrapper.find(PaginationButton).at(1).simulate('click');

      const { currentPage } = mountWrapper.find(Pagination).props();

      expect(currentPage).toEqual(2);
    });
  });
});
