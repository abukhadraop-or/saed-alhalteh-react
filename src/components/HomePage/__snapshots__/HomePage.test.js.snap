// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePage /> tests matches the snapshot 1`] = `
<Fragment>
  <styled.div>
    <styled.h1>
       Conduit 
    </styled.h1>
    <styled.div>
       A place to share your knowledge 
    </styled.div>
  </styled.div>
  <styled.ul>
    <styled.li>
      <Styled(Link)
        to="/"
      >
         Global Feed
      </Styled(Link)>
    </styled.li>
  </styled.ul>
  <styled.div>
    <Tags
      onClick={[Function]}
      selectedTags={Array []}
    />
    <styled.div>
      <Posts
        posts={Array []}
      />
    </styled.div>
  </styled.div>
  <Pagination
    count={0}
    currentPage={1}
    onClick={[Function]}
    pageSize={5}
  />
</Fragment>
`;
